src/source_sender/mod.rs

```diff
@@ -276,4 +321,125 @@ impl Inner {
 
         Ok(())
     }
+
+    /// Calculate the difference between the reference time and the
+    /// timestamp stored in the given event reference, and emit the
+    /// different, as expressed in milliseconds, as a histogram.
+    fn emit_lag_time(&self, event: EventRef<'_>, reference: i64) {
+        if let Some(lag_time_metric) = &self.lag_time {
+            let timestamp = match event {
+                EventRef::Log(log) => log
+                    .get(log_schema().timestamp_key())
+                    .and_then(get_timestamp_millis),
+                EventRef::Metric(metric) => metric
+                    .timestamp()
+                    .map(|timestamp| timestamp.timestamp_millis()),
+                EventRef::Trace(trace) => trace
+                    .get(log_schema().timestamp_key())
+                    .and_then(get_timestamp_millis),
+            };
+            if let Some(timestamp) = timestamp {
+                // This will truncate precision for values larger than 2**52, but at that point the user
+                // probably has much larger problems than precision.
+                let lag_time = (reference - timestamp) as f64 / 1000.0;
```

Comment:
One thing that we may or may not want to do anything about is clock synchronization. If the system clock Vector uses is behind the clock used to generate the event's timestamp, we could end up emitting negative values here.... from the perspective of recording the metrics, we can _handle_ negative values, but they'll look weird downstream.

I'm not actually sure if it makes sense to only allow positive values, or to emit errors when we detect clock desync, or to increment a counter for clock desync errors, etc...

Reply:
Hmm, yes, that it something to highlight. This metric does assume that clocks are accurately synchronized.